File: bookstore/mern/services/category.service.js
import CategoryModel from '../models/category.model.js'

export default class CategoryService {
  async get() {
    return CategoryModel.find({}).lean()
  }

  async getOne(_id) {
    return CategoryModel.findById(_id).lean()
  }

  async create(name) {
    const category = await new CategoryModel({ name })
    await category.save()
    return category
  }

  async change(_id, name) {
    const category = await CategoryModel.findByIdAndUpdate(_id, {
      $set: { name },
    }).lean()
    return { ...category, name }
  }

  async delete(_id) {
    await CategoryModel.findByIdAndRemove(_id)
    return _id
  }
}

File: bookstore/spring-backend/src/main/kotlin/masha/bookstore/services/CategoryService.kt
package masha.bookstore.services

import masha.bookstore.core.BaseService
import masha.bookstore.models.*
import masha.bookstore.interfaces.ICategoryService
import masha.bookstore.repositories.CategoryRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class CategoryService(
    @Autowired override val repository: CategoryRepository
) : BaseService<Category, CategoryRepository>(repository), ICategoryService {

    override fun create(name: String): Category {
        val category = Category(name)
        repository.save(category)
        return category
    }

    override fun change(id: Int, name: String): Category {
        val category: Category = getOne(id)
        category.name = name
        repository.save(category)
        return category
    }

    override fun addBook(book: Book) {
        book.category.let {
            it.books.add(book)
            repository.save(it)
        }
    }

    override fun removeBook(book: Book) {
        book.category.let {
            it.books.remove(book)
            repository.save(it)
        }
    }
}